PHPLIB-881: Consolidate validation of aggregation pipeline options

GromNaN opened the pull request:
Fix PHPLIB-881

Leverage `is_pipeline` to validate `$pipeline` argument and `"pipeline"` option. Since this function returns a boolean, it does not provide context about the error. This change allows to factorise the code and to go further in the validation, but the message in the exception is less precise.

GromNaN commented on src/Operation/Aggregate.php:
```diff
@@ -138,18 +138,9 @@ class Aggregate implements Executable, Explainable
-
-            $expectedIndex += 1;
+        // Empty pipelines are allowed
+        if ($pipeline !== [] && ! is_pipeline($pipeline)) {
```
`is_pipeline` returns `false` for an empty list. There is a test that assert an empty pipeline is valid `AggregateFunctionalTest::testEmptyPipelineReturnsAllDocuments`.

I wonder if empty list is allowed elsewhere and if the constraint should be relaxed in `is_pipeline` to be explicitly stricter when necessary.

alcaeus replied:
I'm on the fence about this. In general I'd say that `is_pipeline` should return `true` for anything that is a valid aggregation pipeline, which includes an empty array. In those cases where we want to forbid using an empty pipeline, we should explicitly guard against a non-empty array. Thoughts @jmikola?

jmikola replied:
Note that `is_pipeline()` was only created for validation of `update` and `findAndModify` command options (not `aggregate`). The original prohibition for empty arrays dates back to https://github.com/mongodb/mongo-php-library/commit/2628c098e3bfc015d7e2d59d1437cf1fed046e12#diff-e093a0fd8f57549d81ad2871e6ca3872ecaef99d0525118e26c0501bdd0c1d92R241 from [PHPLIB-418](https://jira.mongodb.org/browse/PHPLIB-418).

I don't recall why that was done. I recently confirmed in [DRIVERS-2638](https://jira.mongodb.org/browse/DRIVERS-2638) that empty pipelines are accepted for updates.

alcaeus commented on src/Operation/Aggregate.php:
```diff
@@ -138,18 +138,9 @@ class Aggregate implements Executable, Explainable
-            $expectedIndex += 1;
+        // Empty pipelines are allowed
+        if ($pipeline !== [] && ! is_pipeline($pipeline)) {
+            throw new InvalidArgumentException('$pipeline is not a valid list of pipeline operations');
```
```suggestion
            throw new InvalidArgumentException('$pipeline is not a valid aggregation pipeline');
```

alcaeus replied:
Yes, I think that would make sense.

GromNaN replied:
Done.

alcaeus commented on src/Operation/CreateCollection.php:
```diff
@@ -241,20 +241,8 @@ public function __construct(string $databaseName, string $collectionName, array
-                $expectedIndex += 1;
-            }
+        if (isset($options['pipeline']) && ! is_pipeline($options['pipeline'])) {
+            throw new InvalidArgumentException('"pipeline" option is not a valid list of pipeline operations');
```
```suggestion
            throw new InvalidArgumentException('"pipeline" option is not a valid aggregation pipeline');
```

alcaeus reviewed (commented):
Suggested rewording the exception message for clarity, as we generally should refer to this as an "aggregation pipeline" as it is usually called in the documentation.

GromNaN commented on src/Operation/BulkWrite.php:
```diff
@@ -229,7 +229,7 @@ public function __construct(string $databaseName, string $collectionName, array
                     }
 
-                    if (! is_first_key_operator($args[1]) && ! is_pipeline($args[1])) {
+                    if ($args[1] === [] || ! is_first_key_operator($args[1]) && ! is_pipeline($args[1])) {
```
I need to check the spec to see if an empty array is accepted in the following cases:
- BulkWrite UPDATE_MANY and UPDATE_ONE
- FindOneAndUpdate
- ReplaceOne
- Update, UpdateMany, UpdateOne

GromNaN replied:
So I'll change the behavior for `ReplaceOne` that is currently rejecting empty array.

alcaeus replied:
Should be fine - just make sure it is covered by a test.

FWIW, the shell behaves the same way:
```
test> // Replace with empty replacement
test> db.test.replaceOne({_id: 1}, {})
{
  acknowledged: true,
  insertedId: null,
  matchedCount: 1,
  modifiedCount: 1,
  upsertedCount: 0
}
test> // Update with empty document
test> db.test.updateOne({_id: 1}, {})
MongoInvalidArgumentError: Update document requires atomic operators
test> // Update with empty pipeline
test> db.test.updateOne({_id: 1}, [])
MongoInvalidArgumentError: Update document requires atomic operators
```

GromNaN replied:
Thanks for checking, I'll adopt this reflex.

Test case already added: https://github.com/mongodb/mongo-php-library/pull/1078/files#diff-6e13b3e2ae70bee5981233671f8a994c0111a3cf84e2f958e723bdd08c9c83d2R45

jmikola commented on tests/Operation/ReplaceOneTest.php:
```diff
@@ -31,29 +31,31 @@ public function testConstructorReplacementArgument($replacement): void
     }
 
-    public function provideUpdateDocuments()
+    public function provideInvalidReplacementDocuments()
```
Note: this is going to conflict with https://github.com/mongodb/mongo-php-library/pull/1077. The changes in that PR are more involved, so I'd suggest waiting on that to be merged.

Consideration of PackedArray instances for pipeline values is also going to impact this PR.

GromNaN commented:
PR rebased.

> Consideration of PackedArray instances for pipeline values is also going to impact this PR.

The dumb comparison `$pipeline === []` is not possible with `PackedArray`. So I added a, argument to `is_pipeline` to specify the return value in case of empty list. This also simplified conditions in existing `if` statements.

alcaeus reviewed (approved):
LGTM. While I'm not a fan of the comments indicating an option name, I'm fine with having them since we can't use named arguments for the purpose.

alcaeus commented:
> I copied the style introduced by @jmikola and [discussed in review](https://github.com/mongodb/mongo-php-library/pull/1098/files#r1223980183). Could be reverted.

Ah, I missed that one. No worries then, perfectly fine with it 👍
